File: rules.py
"""Rule engine: YAML pack loader + simpleeval interpreter (PRD §7).

Rules are data, not code (FR-10). A domain expert edits the YAML packs;
this module only knows how to evaluate `logic` expressions.

FR-9: a rule with a missing/unconfirmed required input emits a
"cannot evaluate" finding — never a silent pass.
Grill-me decision: verify_flag rules cap at needs_verification.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

import yaml
from simpleeval import EvalWithCompoundTypes

from report import ExtractedParam, Finding

PACK_DIR = Path(__file__).resolve().parent / "rule_packs"

_REGIME = {"BNBC-2020": "BNBC", "RAJUK-DAP/Bidhimala": "RAJUK"}


@dataclass
class Rule:
    id: str
    source: str
    title: str
    parameters: list
    logic: str
    severity: str
    confidence_basis: str
    citation: str
    remediation: str
    verify_flag: bool = False


def load_packs(pack_dir: Path = PACK_DIR) -> list[Rule]:
    rules = []
    for path in sorted(pack_dir.glob("*.yaml")):
        pack = yaml.safe_load(path.read_text())
        for row in pack["rules"]:
            rules.append(Rule(source=pack["source"], **row))
    return rules


def _consistency(values, tolerance=0.05):
    """True when all values agree within relative tolerance."""
    nums = [float(v) for v in values]
    lo, hi = min(nums), max(nums)
    return hi == 0 or (hi - lo) / max(abs(hi), 1e-9) <= tolerance


def evaluate_rules(rules: list[Rule], params: list[ExtractedParam]) -> list[Finding]:
    confirmed = {p.param: p.value for p in params if p.confirmed}
    crops = {p.param: p.source_crop for p in params if p.confirmed}

    findings = []
    for rule in rules:
        missing = [name for name in rule.parameters if name not in confirmed]
        regime = _REGIME[rule.source]
        crop = next((crops[n] for n in rule.parameters if crops.get(n)), None)

        if missing:
            findings.append(Finding(
                rule_id=rule.id, bucket="needs_verification",
                severity=rule.severity, confidence=0.0,
                reason=(f"Cannot evaluate — input missing or unconfirmed: "
                        f"{', '.join(missing)}."),
                citation=rule.citation, regime=regime,
                inputs_used={k: confirmed[k] for k in rule.parameters
                             if k in confirmed},
                sheet_location=crop, remediation=rule.remediation,
                verify_flag=rule.verify_flag,
            ))
            continue

        names = dict(confirmed)
        evaluator = EvalWithCompoundTypes(
            names=names,
            functions={
                "param_present": lambda n, _c=confirmed: n in _c and bool(_c[n]),
                "consistency": _consistency,
            },
        )
        passed = bool(evaluator.eval(rule.logic))
        inputs_used = {k: confirmed[k] for k in rule.parameters}

        if passed:
            bucket, reason = "appears_compliant", (
                f"{rule.title}: appears compliant on confirmed inputs "
                f"(not a certification)."
            )
        else:
            bucket = "needs_verification" if rule.verify_flag else "likely_violation"
            reason = f"{rule.title}: check failed on confirmed inputs."
            if rule.verify_flag:
                reason += (" Threshold is [VERIFY] (2025/2026 rules in flux) — "
                           "expert must confirm the value before relying on this.")

        findings.append(Finding(
            rule_id=rule.id, bucket=bucket, severity=rule.severity,
            confidence=min((p.confidence for p in params
                            if p.confirmed and p.param in rule.parameters),
                           default=1.0),
            reason=reason, citation=rule.citation, regime=regime,
            inputs_used=inputs_used, sheet_location=crop,
            remediation=rule.remediation, verify_flag=rule.verify_flag,
        ))
    return findings

File: web/src/engine/rules.ts
/**
 * Rule engine — TS port of rules.py with identical semantics (PRD §7).
 *
 * Rules are data, not code (FR-10): the YAML packs are shared byte-for-byte
 * with the Python engine. `logic` strings are written in the packs'
 * Python-ish grammar; a tiny transpile step maps them to JS before eval.
 *
 * FR-9: a rule with a missing/unconfirmed required input emits a
 * "cannot evaluate" finding — never a silent pass.
 * verify_flag rules cap at needs_verification.
 */
import { load as parseYaml } from 'js-yaml'

export interface ExtractedParam {
  param: string
  value: unknown
  unit: string
  confidence: number
  sourcePage: number
  sourceCrop: string | null // data-URL crop image
  cropFallbackNote?: string | null
  bbox?: [number, number, number, number] | null // page-pixel coords
  confirmed: boolean
  editedFrom?: unknown
}

export interface Rule {
  id: string
  source: string
  title: string
  parameters: string[]
  logic: string
  severity: string
  confidenceBasis: string
  citation: string
  remediation: string
  verifyFlag: boolean
}

export type Bucket = 'likely_violation' | 'needs_verification' | 'appears_compliant'

export interface Finding {
  ruleId: string
  bucket: Bucket
  severity: string
  confidence: number
  reason: string
  citation: string
  regime: 'BNBC' | 'RAJUK'
  inputsUsed: Record<string, unknown>
  sheetLocation: string | null
  remediation: string
  verifyFlag: boolean
  userAction?: 'accept' | 'dismiss' | null
  userNote?: string | null
}

const REGIME: Record<string, 'BNBC' | 'RAJUK'> = {
  'BNBC-2020': 'BNBC',
  'RAJUK-DAP/Bidhimala': 'RAJUK',
}

interface PackYaml {
  source: string
  rules: Array<{
    id: string
    title: string
    parameters: string[]
    logic: string
    severity: string
    confidence_basis: string
    citation: string
    remediation: string
    verify_flag?: boolean
  }>
}

export function parsePack(yamlText: string): Rule[] {
  const pack = parseYaml(yamlText) as PackYaml
  return pack.rules.map((row) => ({
    id: row.id,
    source: pack.source,
    title: row.title,
    parameters: row.parameters,
    logic: row.logic,
    severity: row.severity,
    confidenceBasis: row.confidence_basis,
    citation: row.citation,
    remediation: row.remediation,
    verifyFlag: row.verify_flag ?? false,
  }))
}

/**
 * Python-ish → JS. Handles exactly the grammar the packs use:
 * one-level `X if C else Y`, `and/or/not`, `True/False`, list literals
 * (same syntax), and helper calls. Packs are repo-controlled data — same
 * trust level as code.
 */
export function transpileLogic(logic: string): string {
  let js = logic
    .replace(/\bTrue\b/g, 'true')
    .replace(/\bFalse\b/g, 'false')
    .replace(/\bnot\s+/g, '!')
    .replace(/\band\b/g, '&&')
    .replace(/\bor\b/g, '||')
  // single conditional expression: X if C else Y  →  (C) ? (X) : (Y)
  const cond = js.match(/^(.*?)\s+if\s+(.*?)\s+else\s+(.*)$/)
  if (cond) js = `(${cond[2]}) ? (${cond[1]}) : (${cond[3]})`
  return js
}

function consistency(values: unknown[], tolerance = 0.05): boolean {
  const nums = values.map(Number)
  const lo = Math.min(...nums)
  const hi = Math.max(...nums)
  return hi === 0 || (hi - lo) / Math.max(Math.abs(hi), 1e-9) <= tolerance
}

function evalLogic(logic: string, names: Record<string, unknown>): boolean {
  const scope = {
    ...names,
    param_present: (n: string) => n in names && Boolean(names[n]),
    consistency,
  }
  // non-strict Function body so `with` is allowed; packs are trusted repo data
  const fn = new Function('scope', `with (scope) { return (${transpileLogic(logic)}) }`)
  return Boolean(fn(scope))
}

export function evaluateRules(rules: Rule[], params: ExtractedParam[]): Finding[] {
  const confirmed: Record<string, unknown> = {}
  const crops: Record<string, string | null> = {}
  for (const p of params) {
    if (p.confirmed) {
      confirmed[p.param] = p.value
      if (p.sourceCrop) crops[p.param] = p.sourceCrop
    }
  }

  return rules.map((rule) => {
    const regime = REGIME[rule.source]
    const missing = rule.parameters.filter((n) => !(n in confirmed))
    const crop = rule.parameters.map((n) => crops[n]).find(Boolean) ?? null

    if (missing.length) {
      return {
        ruleId: rule.id,
        bucket: 'needs_verification' as const,
        severity: rule.severity,
        confidence: 0.0,
        reason: `Cannot evaluate — input missing or unconfirmed: ${missing.join(', ')}.`,
        citation: rule.citation,
        regime,
        inputsUsed: Object.fromEntries(
          rule.parameters.filter((n) => n in confirmed).map((n) => [n, confirmed[n]]),
        ),
        sheetLocation: crop,
        remediation: rule.remediation,
        verifyFlag: rule.verifyFlag,
      }
    }

    const passed = evalLogic(rule.logic, confirmed)
    const inputsUsed = Object.fromEntries(rule.parameters.map((n) => [n, confirmed[n]]))

    let bucket: Bucket
    let reason: string
    if (passed) {
      bucket = 'appears_compliant'
      reason = `${rule.title}: appears compliant on confirmed inputs (not a certification).`
    } else {
      bucket = rule.verifyFlag ? 'needs_verification' : 'likely_violation'
      reason = `${rule.title}: check failed on confirmed inputs.`
      if (rule.verifyFlag) {
        reason +=
          ' Threshold is [VERIFY] (2025/2026 rules in flux) — expert must confirm the value before relying on this.'
      }
    }

    const confidences = params
      .filter((p) => p.confirmed && rule.parameters.includes(p.param))
      .map((p) => p.confidence)

    return {
      ruleId: rule.id,
      bucket,
      severity: rule.severity,
      confidence: confidences.length ? Math.min(...confidences) : 1.0,
      reason,
      citation: rule.citation,
      regime,
      inputsUsed,
      sheetLocation: crop,
      remediation: rule.remediation,
      verifyFlag: rule.verifyFlag,
    }
  })
}
